app/controllers/application_controller.rb

```diff
@@ -5,6 +5,140 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
+  get '/' do
+    @user = User.find(session[:user_id]) if session[:user_id]
+    erb :index
+  end
+
+  get "/signup" do
+    if !logged_in?
+      erb :'/users/signup'
+    else
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/signup" do
+    user = User.new(username: params[:username], email: params[:email], password: params[:password])
+    if user.save
+      session[:user_id] = user.id
+      redirect '/tweets', locals: {message: "Sign up successful, please log in now."}
+    else
+      redirect '/signup', locals: {message: "Sign up failed, try again."}
+    end
+  end
+
+  get "/login" do
+    if !logged_in?
+      erb :'/users/login'
+    else
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets", locals: {message: "Successfully logged in."}
+    else
+      redirect "/login", locals: {message: "Log in failed, try again."}
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.clear
+      redirect "/login", locals: {message: "Successfully logged out."}
+    else
+      redirect "/login", locals: {message: "Can't log out, you're not logged in."}
+    end
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get "/tweets" do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).